[Payment Request] Add support for merchant category codes.

MiaMukherjee opened the pull request:
#### 65e05b1cdf9b5b3b6262d9713d82ddc85eba1bba
<pre>
[Payment Request] Add support for merchant category codes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270978">https://bugs.webkit.org/show_bug.cgi?id=270978</a>
&lt;<a href="https://rdar.apple.com/problem/124266155">rdar://problem/124266155</a>&gt;

Reviewed by Abrar Rahman Protyasha.

This patch adds support for merchant category codes and populates the relevant field in `PKPaymentRequest`.

Tests: http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayRequestBase.h:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.idl:
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
(WebCore::ApplePaySessionPaymentRequest::merchantCategoryCode const):
(WebCore::ApplePaySessionPaymentRequest::setMerchantCategoryCode):
(WebCore::ApplePaySessionPaymentRequest::ApplePaySessionPaymentRequest):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
* Source/WebCore/testing/MockPaymentCoordinator.h:
* Source/WebCore/testing/MockPaymentCoordinator.idl:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::toPKMerchantCategoryCode):
(WebKit::WebPaymentCoordinatorProxy::platformPaymentRequest):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https-expected.txt
* LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html

* LayoutTests/platform/ios-wk2/TestExpectations
* LayoutTests/platform/mac-wk2/TestExpectations

Canonical link: <a href="https://commits.webkit.org/278718@main">https://commits.webkit.org/278718@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b86617dffc44c891217c41b5d5460e3059e9b6db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30643 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54600 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2027 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1706 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41814 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53436 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28289 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22932 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25611 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1527 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44678 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47581 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1600 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56190 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50839 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26452 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1495 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49207 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27695 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44332 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48384 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11242 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28585 "Built successfully") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/63163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27430 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/63163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,58 @@
+    }, "Should have a feature for `paymentRequestMerchantCategoryCode`.");
+    
+    user_activation_test(async (test) => {
+        let request = new PaymentRequest([validPaymentMethod()], validPaymentDetails());
```
```suggestion
        const request = new PaymentRequest([validPaymentMethod()], validPaymentDetails());
```

Same with the other `let`s, unless they are being reassigned.

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,58 @@
+            internals.mockPaymentCoordinator.acceptPayment();
+        });
+        
+        let response = await request.show();
```
I wonder if you should wrap this in a promise to assure sure it's actually getting run before the test finishes. I would do something like: 

```suggestion
        const shippingPromise = new Promise(resolve => {
            request.addEventListener("shippingaddresschange", (event) => {
                assert_equals(internals.mockPaymentCoordinator.merchantCategoryCode, null, "check that the `merchantCategoryCode` is still `null` after an update");
                event.updateWith({ });
                internals.mockPaymentCoordinator.acceptPayment();
                resolve();
            });
        });

        const response = await request.show();
        await shippingPromise;
```

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,58 @@
+        request.addEventListener("merchantvalidation", (event) => {
+            event.complete({ });
+        });
+        request.addEventListener("shippingaddresschange", (event) => {
```
As above... might be worth wrapping in a promise.

MiaMukherjee replied:
As I mentioned below, the `event.updateWith` function doesn't like being called from an async context, so I'm going to leave Promises alone for this change.

marcoscaceres reviewed (commented):
Couple of suggestions

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,76 @@
+<body>
+<script>
+    setup({ explicit_done: true, explicit_timeout: true });
+    console.log("Bleh");
```
I think you might have forgotten to remove these?

MiaMukherjee replied:
Oh dear 😅 Definitely didn't intend to include those, thanks for pointing them out!

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,76 @@
+    setup({ explicit_done: true, explicit_timeout: true });
+    console.log("Bleh");
+    test((test) => {
+        console.log("Blah");
```
same here...

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,73 @@
+        
+        assert_equals(internals.mockPaymentCoordinator.merchantCategoryCode, null, "check that the `merchantCategoryCode` is still `null` after the payment is accepted");
+        
+        await response.complete("success");
```
The test doesn't seem to match the description? Shouldn't it be: 

```suggestion
        const shippingPromise = new Promise((resolve) => {
            request.addEventListener("shippingaddresschange", resolve);
        });
        const response = await request.show();
        
        const event = await shippingPromise;
        event.updateWith({});
        assert_equals(
            internals.mockPaymentCoordinator.merchantCategoryCode,
            null,
            "check that the `merchantCategoryCode` is still `null` after an update"
        );
        
        internals.mockPaymentCoordinator.acceptPayment();
        assert_equals(
            internals.mockPaymentCoordinator.merchantCategoryCode,
            null,
            "check that the `merchantCategoryCode` is still `null` after the payment is accepted"
        );
        
        await response.complete("success");

```

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,73 @@
+    
+    user_activation_test(async (test) => {
+        const method = validPaymentMethod();
+        method.data.merchantCategoryCode = "0739";
```
It might be good to have a comment as to what "0739" means (if anything?). And maybe a link to where it's defined if it's meaningful.

marcoscaceres replied:
Also, what happens if one provides a bogus method.data.merchantCategoryCode? would that reject? Should there be a test for that?

MiaMukherjee replied:
The specification for MCCs doesn't actually say what their value should be, it's only through convention that they're four-digit numerical codes. Because of that, we want to accept all strings on the API level and only convert them down to numbers when they get passed to the PassKit payment request.

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,73 @@
+        await response.complete("success");
+    }, "Should propagate `merchantCategoryCode` if provided.");
+    
+    done();
```
Is this done() needed?

MiaMukherjee replied:
I don't know. I inferred it was necessary from the other tests in the project, but I'm not sure what it actually does here.

marcoscaceres replied:
Yeah, try removing it. I don't think it does anything.

MiaMukherjee replied:
If it's removed the tests time out

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html:
```diff
@@ -0,0 +1,73 @@
+        
+        assert_equals(internals.mockPaymentCoordinator.merchantCategoryCode, method.data.merchantCategoryCode, "check that the `merchantCategoryCode` still matches after the payment is accepted");
+        
+        await response.complete("success");
```
Maybe, just: 

```suggestion
        const shippingPromise = new Promise(resolve => {
            request.addEventListener("shippingaddresschange", resolve);
        });
        
        const response = await request.show();
        
        const event = await shippingPromise;
        assert_equals(
          internals.mockPaymentCoordinator.merchantCategoryCode,
          method.data.merchantCategoryCode, 
          "check that the `merchantCategoryCode` still matches after an update"
        );
        event.updateWith({ });
        
        internals.mockPaymentCoordinator.acceptPayment();
        assert_equals(
          internals.mockPaymentCoordinator.merchantCategoryCode, 
          method.data.merchantCategoryCode,
          "check that the `merchantCategoryCode` still matches after the payment is accepted"
        );
        
        await response.complete("success");
```

MiaMukherjee replied:
So I tried using promises, but that causes trouble with the payment request internals. Unfortunately it expects the `updateWith` call to be done synchronously and so doesn't play nice with Promises.

I think it's ok to skip using them here, especially as none of the other payment request tests use it and the consistency might help readability for someone looking through.

marcoscaceres reviewed (commented):
Just some more minor things with the tests

aprotyas commented on Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
```diff
@@ -179,6 +179,13 @@ class ApplePaySessionPaymentRequest {
+    const String& merchantCategoryCode() const { return m_merchantCategoryCode; }
+    void setMerchantCategoryCode(const String& merchantCategoryCode)
+    {
+        m_merchantCategoryCode = merchantCategoryCode; }
```
```suggestion
        m_merchantCategoryCode = merchantCategoryCode;
    }
```

MiaMukherjee replied:
Instead of dropping the `}` to the next line I included the whole function in the same line to match the others in the header.

aprotyas commented on Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
```diff
@@ -259,6 +269,9 @@ class ApplePaySessionPaymentRequest {
             , m_applePayLaterAvailability(WTFMove(applePayLaterAvailability))
+#endif
+#if ENABLE(APPLE_PAY_MERCHANT_CATEGORY_CODE)
+            , m_merchantCategoryCode(WTFMove(merchantCategoryCode))
```
Hmm, there's no point in moving an lvalue. Consider making `merchantCategoryCode` an rvalue reference to match the rest of the arguments.

aprotyas reviewed (commented):
A couple of comments!

MiaMukherjee committed:
[Payment Request] Add support for merchant category codes.
https://bugs.webkit.org/show_bug.cgi?id=270978
<rdar://problem/124266155>

Reviewed by Abrar Rahman Protyasha.

This patch adds support for merchant category codes and populates the relevant field in `PKPaymentRequest`.

Tests: http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayRequestBase.h:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.idl:
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
(WebCore::ApplePaySessionPaymentRequest::merchantCategoryCode const):
(WebCore::ApplePaySessionPaymentRequest::setMerchantCategoryCode):
(WebCore::ApplePaySessionPaymentRequest::ApplePaySessionPaymentRequest):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
* Source/WebCore/testing/MockPaymentCoordinator.h:
* Source/WebCore/testing/MockPaymentCoordinator.idl:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::toPKMerchantCategoryCode):
(WebKit::WebPaymentCoordinatorProxy::platformPaymentRequest):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https-expected.txt
* LayoutTests/http/tests/paymentrequest/paymentrequest-merchantCategoryCode.https.html

* LayoutTests/platform/ios-wk2/TestExpectations
* LayoutTests/platform/mac-wk2/TestExpectations

Canonical link: https://commits.webkit.org/278718@main
